### ISSUE TITLE
Incorrect behavior when using alias refers to namespace in using alias

### ISSUE THREAD

**Issue by amcasey** [dotnet/roslyn#1027]
``` C#
extern alias Contracts1;
extern alias Contracts2;

using alias1 = Contracts1::TestContractAssembly;
using alias2 = Contracts2::TestContractAssembly;

partial class Contract : alias1.IContract, alias2.IContract
{
    int alias1.IContract.x
    {
        get
        {
            return 1;
        }
    }
    int alias2.IContract.x
    {
        get
        {
            return 1;
        }
    }
}
```

Both IContract types are the same - they expose an int property "x".  Roslyn incorrectly reports that IContract.x has already been implemented.

This works if you inline the using aliases.

I strongly suspect that the alias points to the merged namespace symbol, rather than to the namespace symbol contained in the assembly symbol referenced by the extern alias.

**Comment by theoy** [comment]
Note, this was reported by a Microsoft team based on their real world source code, in their migration to use Roslyn.

**Comment by VladimirReshetnikov** [comment]
@amcasey @VSadov @theoy My experiments show that Dev12 compiles this code without errors, but metadata for class `Contract` contains 2 properties with identical signatures `int32 TestContractAssembly.IContract.x()` - it fails to verify and fails to load at runtime. Do we have more info about the real repro? Was it a usable assembly?

If we inline aliases, then the C# compiler includes the extern alias as part of the property name, and it works ok. In principle, we could always generate unique names, but it would require more work.

**Comment by VladimirReshetnikov** [comment]
@amcasey So far my tests show that aliases correctly refer to distinct non-merged namespaces with `NamespaceKind.Module`.

**Comment by amcasey** [comment]
The repro was reduced from an internal partner's code.  I believe they care about the aliases referring to different namespace constituents but don't actually try to implement two different copies of the same interface on the same type.

**Comment by amcasey** [comment]
@VladimirReshetnikov Then why are we getting errors about implementing the same property twice?

**Comment by VladimirReshetnikov** [comment]
I'm going to close it as "By Design" for now. Let me know if you have any reasons to believe it should compile successfully, and what is the expected output.

**Comment by amcasey** [comment]
@VladimirReshetnikov explained what I was missing offline: the diagnostic indicates that two names are colliding, not that we're implementing the same interface property twice.  I agree that the behavior is reasonable (if suboptimal) and not a regression.

We'll follow up offline with the team that originally reported the issue.

**Comment by gafter** [comment]
I don't see how this can be by design. The language specification says clearly that this should work, and it does not work. I understand that it is not a regression because the native compiler could not handle this, but it is still a bug. What am I missing?

**Comment by VladimirReshetnikov** [comment]
The issue does not say what is the expected behavior (including what metadata is expected to be emitted).

**Comment by gafter** [comment]
No, the issue does not say. But the language clearly defines what should happen, no?

**Comment by VladimirReshetnikov** [comment]
No, the spec is silent about what names in metadata are given to explicit interface implementations. We have established rules for this (it might be breaking to change them), but unfortunately these rules sometimes result in several members of the class having identical names. The underlying platform treats this situation as an error and refuse to load such a type. The native C# compiler did not report any error, and left the problem to be discovered at runtime. In Roslyn we decided to give a compile-time error in such cases.

The repro steps incorrectly state that Roslyn reports that IContract.x has already been implemented. In fact, there is a different error `error CS0102: The type 'Contract' already contains a definition for 'TestContractAssembly.IContract.x'`. It just indicates a name collision.

**Comment by gafter** [comment]
Of course the spec doesn't say what names are given to explicit implementations in metadata, and neither does this issue request particular metadata names. But the language spec does say that this program should be accepted and defines its semantics.

Our rules never result in several members of the class having identical names *in programs that compile today*, so making programs such as this one compile correctly need not require any breaking changes.

**Comment by gafter** [comment]
Having said that, this probably doesn't need to be fixed in the current release.

**Comment by VladimirReshetnikov** [comment]
Suppose we decide to always inlude an extern alias in the member name, even if it indirectly referenced by a using alias. It would change metadata names in some programs that were successfuly compiler with the native compiler. If the program uses reflection to access those members, it could be broken. Or, adding an alias to an already long name could overflow name length limitation in CLR. All these cases need to be carefully considered before we decide to improve Roslyn compiler to successfully compile new programs.

**Comment by gafter** [comment]
@VladimirReshetnikov I agree the proposed fixes you describe could break some clients. That suggests we need a subtler change, perhaps only in cases where there are conflicts.

**Comment by jaredpar** [comment]
Based on this being ruled as "By Design" spec wise and having no other significant report for all of Roslyn 1.X I'm going to close this issue out.

**Comment by gafter** [comment]
This is not "By Design" spec-wise; the language spec directs that the compiler accept the program and produce working code for it. The failure is a result of the *compiler design*, which results in either bad metadata or an error message. See https://github.com/dotnet/roslyn/issues/1027#issuecomment-78404658 for details. The language spec does not require that we retain that design error in the compiler, so I'm reopening.
